rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,541 @@
+
+- Configurable cluster alerts when the number of available renewals has crossed
+  some threshold
+
+- Exposing the number of available renewals in the web UI and `tctl bot ls`
+
+- Exposing per-token renewal counts as Prometheus metrics, both on the Auth
+  Service and via `tbot`'s metrics endpoint.
+
+#### Outstanding Issue: Soft Bot Expiration
+
+The `spec.rejoining.expires` field can be used to prevent rejoining after a
+certain time, in tandem with the rejoin count limit. This has the - likely
+confusing - downside that bots will still be able to renew their certificates
+indefinitely past the expiration date, assuming their certs were valid at the
+time of expiration.
+
+Also, as with all Teleport resources, the `metadata.expires` field can also
+remove the token resource after a set time. Bots will also continue to renew
+certs as long as possible until they are either locked or otherwise fail to
+renew their certs on time.
+
+These two expirations create some confusion, and do not allow for an obvious
+method to deny a bot access, aside from creating a lock.
+
+We would like to solve two expiration use cases:
+1. We should be able to prevent all bot resource access after a certain date,
+   including renewals, in a way that allows the bot to be resumed later if
+   desired. (I.e. the token resource must still exist.)
+
+   Locks may accomplish this, but some centralized management in the token
+   resource would be convenient.
+
+2. We should be able to prevent bot rejoins after a certain date, to control
+   rejoining conditions in tandem with the rejoin counter. The
+   `spec.rejoining.expires` field accomplishes this, but does have a naming
+   collision with `metadata.expires`.
+
+TODO: Ensure this is solved and not confusing.
+
+#### Keypair Rotation
+
+Given the long-lived nature of the keypair credential, it's important to support
+rotation without bot downtime. Ideally, it should be possible to initiate a
+rotation from either the server (e.g. by setting a `rotate_on_next_renewal` flag
+on the token/bot instance) or `tbot` client.
+
+TODO: Expand this.
+
+#### Remaining Downsides
+
+- Repairing a bot that has exhausted all of its rejoins is still a semi-manual
+  process. It is significantly easier, and does not necessarily require any
+  changes on the impacted bot node itself, but is still annoying. Users can opt
+  out of this by setting `.spec.rejoining.unlimited=true`, but this has obvious
+  security implications.
+
+- Effort required to configure IaC / Terraform is still fairly high, even if
+  reduced.
+
+### Other Supporting Improvements
+
+Alongside `challenge` joining, we have several UX proposals to further improve
+the usability of non-delegated joining.
+
+#### Longer-Lived Bots
+
+The renewable identity's 24 hour maximum TTL is too restrictive and should be
+lengthened. We propose raising this limit to 7 days, but keeping the default
+(1hr) the same.
+
+#### Bot Instance Locking
```

Comment:
Do we also want the ability to put a lock on the public key / join token to prevent compromised bots from getting new credentials?

Reply:
Yes, it will (at least) be indirectly linked to a join token via the bot instance, but locking on public key could work too. I'll try to evaluate which avenue will work best, or maybe all if it's relatively simple to add the additional lock targets.

Reply:
I've added some details on specific new lock targets, including public key.